File: contracts/rust/tokenregistry/tokenregistry.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package tokenregistry

import "github.com/iotaledger/wasp/packages/vm/wasmlib"

func funcMintSupply(ctx *wasmlib.ScFuncContext, params *FuncMintSupplyParams) {
	minted := ctx.Incoming().Minted()
	if minted.Equals(wasmlib.MINT) {
		ctx.Panic("TokenRegistry: No newly minted tokens found")
	}
	state := ctx.State()
	registry := state.GetMap(VarRegistry).GetBytes(minted)
	if registry.Exists() {
		ctx.Panic("TokenRegistry: Color already exists")
	}
	token := &Token{
		Supply:      ctx.Incoming().Balance(minted),
		MintedBy:    ctx.Caller(),
		Owner:       ctx.Caller(),
		Created:     ctx.Timestamp(),
		Updated:     ctx.Timestamp(),
		Description: params.Description.Value(),
		UserDefined: params.UserDefined.Value(),
	}
	if token.Supply <= 0 {
		ctx.Panic("TokenRegistry: Insufficient supply")
	}
	if len(token.Description) == 0 {
		token.Description += "no dscr"
	}
	registry.SetValue(token.Bytes())
	colors := state.GetColorArray(VarColorList)
	colors.GetColor(colors.Length()).SetValue(minted)
}

func funcTransferOwnership(_sc *wasmlib.ScFuncContext, params *FuncTransferOwnershipParams) {
	//TODO
}

func funcUpdateMetadata(_sc *wasmlib.ScFuncContext, params *FuncUpdateMetadataParams) {
	//TODO
}

func viewGetInfo(_sc *wasmlib.ScViewContext, params *ViewGetInfoParams) {
	//TODO
}

File: contracts/rust/tokenregistry/src/tokenregistry.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use wasmlib::*;

use crate::*;
use crate::types::*;

pub fn func_mint_supply(ctx: &ScFuncContext) {
    ctx.log("tokenregistry.mintSupply");
    let p = ctx.params();
    let param_description = p.get_string(PARAM_DESCRIPTION);
    let param_user_defined = p.get_string(PARAM_USER_DEFINED);

    let minted = ctx.incoming().minted();
    if minted == ScColor::MINT {
        ctx.panic("TokenRegistry: No newly minted tokens found");
    }
    let state = ctx.state();
    let registry = state.get_map(VAR_REGISTRY).get_bytes(&minted);
    if registry.exists() {
        ctx.panic("TokenRegistry: Color already exists");
    }
    let mut token = Token {
        supply: ctx.incoming().balance(&minted),
        minted_by: ctx.caller(),
        owner: ctx.caller(),
        created: ctx.timestamp(),
        updated: ctx.timestamp(),
        description: param_description.value(),
        user_defined: param_user_defined.value(),
    };
    if token.supply <= 0 {
        ctx.panic("TokenRegistry: Insufficient supply");
    }
    if token.description.is_empty() {
        token.description += "no dscr";
    }
    registry.set_value(&token.to_bytes());
    let colors = state.get_color_array(VAR_COLOR_LIST);
    colors.get_color(colors.length()).set_value(&minted);
    ctx.log("tokenregistry.mintSupply ok");
}

pub fn func_transfer_ownership(ctx: &ScFuncContext) {
    ctx.log("tokenregistry.transferOwnership");
    //TODO the one who can transfer token ownership
    ctx.require(ctx.caller() == ctx.contract_creator(), "no permission");

    let p = ctx.params();
    let param_color = p.get_color(PARAM_COLOR);
    ctx.require(param_color.exists(), "missing mandatory color");

    //TODO
    ctx.log("tokenregistry.transferOwnership ok");
}

pub fn func_update_metadata(ctx: &ScFuncContext) {
    ctx.log("tokenregistry.updateMetadata");
    //TODO the one who can change the token info
    ctx.require(ctx.caller() == ctx.contract_creator(), "no permission");

    let p = ctx.params();
    let param_color = p.get_color(PARAM_COLOR);
    ctx.require(param_color.exists(), "missing mandatory color");

    //TODO
    ctx.log("tokenregistry.updateMetadata ok");
}

pub fn view_get_info(ctx: &ScViewContext) {
    ctx.log("tokenregistry.getInfo");
    let p = ctx.params();
    let param_color = p.get_color(PARAM_COLOR);
    ctx.require(param_color.exists(), "missing mandatory color");

    //TODO
    ctx.log("tokenregistry.getInfo ok");
}
